tensorbay/cli/log.py

```diff
@@ -158,10 +166,225 @@ def _get_full_log(commit: Commit, branch_name: Optional[str]) -> str:
+            # Save commit to commit dict.
+            commit_to_node[commit_id] = current_node
+            child_node = current_node
+
+    # Check the correction of leaf set.
+    for commit_id in tuple(leaves):
+        if leaves[commit_id].available_child_num != 0:
+            del leaves[commit_id]
+    return sorted(leaves.values(), key=lambda x: x.commit.committer.date, reverse=True)
+
+
+class _GraphPrinter:  # pylint: disable=too-few-public-methods
+    """This class defines the structure of logging graphical commits stack.
+
+    Arguments:
+        sorted_leaves: The leaves of commit tree sorted by date.
+        oneline: The log method.
+
+    """
+
+    def __init__(
+        self,
+        sorted_leaves: List[_CommitNode],
+        oneline: bool,
+        commit_id_to_branches: Dict[str, List[str]],
+    ):
+        self._sorted_leaves = sorted_leaves
+        self._layer = 1
+        self._pointer = 0
+        self._source_pointer: Optional[int] = None
+        self._target_pointer = 0
+        self._commit_id_to_branches = commit_id_to_branches
+
+        if oneline:
+            self._graph_printer = partial(self._add_graph_oneline, printer=_get_oneline_log)
+        else:
+            self._graph_printer = partial(self._add_graph_full, printer=_get_full_log)
+
+    def _get_log_node(self) -> _CommitNode:
+        """Get the next log commit node.
+
+        Returns:
+            The next log commit node.
+
+        Raises:
+            RuntimeError: Graphic logging algorithm error.
+
+        """
+        self._source_pointer = None
+        for log_node in islice(self._sorted_leaves, self._pointer, None):
+            if log_node.available_child_num == 0:
+                self._target_pointer = self._pointer
+                return log_node
+
+            self._pointer += 1
+            if self._pointer >= self._layer:
+                self._layer += 1
+        raise RuntimeError("Graphic logging algorithm error.")
+
+    def _merge_branches(self, parent: Optional[_CommitNode]) -> None:
+        """Merge branches.
+
+        Arguments:
+            parent: The parent node.
+
+        """
+        if parent not in self._sorted_leaves:
+            return
+        index = self._sorted_leaves.index(parent)
+        self._target_pointer, self._source_pointer = sorted((index, self._pointer))
+        self._layer -= 1
+        del self._sorted_leaves[self._source_pointer]
+
+    def _set_next_node(self, node: _CommitNode) -> None:
+        """Set the next node at the position of the printing pointer.
+
+        Arguments:
+            node: The next node of the printing pointer.
+
+        """
+        node.available_child_num -= 1
+        self._pointer = self._target_pointer
+        self._sorted_leaves[self._pointer] = node
+
+    def _add_graph_oneline(
+        self,
+        commit: Commit,
+        branch_name: Optional[str],
+        *,
+        printer: Callable[[Commit, Optional[str]], str],
```

Comment:
Is it possible to remove the argument `printer` here, and use `_get_oneline_log` in the method directly?
Same in `_add_graph_full`.

Reply:
OK~ Madam~